src/textual/_compositor.py

```diff
@@ -0,0 +1,549 @@
+from .geometry import Region, Offset, Size
+
+
+from ._loop import loop_last
+from ._types import Lines
+from .widget import Widget
+
+if sys.version_info >= (3, 10):
+    from typing import TypeAlias
+else:  # pragma: no cover
+    from typing_extensions import TypeAlias
+
+
+if TYPE_CHECKING:
+    from .screen import Screen
+    from .widget import Widget
+
+
+class ReflowResult(NamedTuple):
+    """The result of a reflow operation. Describes the chances to widgets."""
+
+    hidden: set[Widget]  # Widgets that are hidden
+    shown: set[Widget]  # Widgets that are shown
+    resized: set[Widget]  # Widgets that have been resized
+
+
+class RenderRegion(NamedTuple):
+    """Defines the absolute location of a Widget."""
+
+    region: Region  # The region occupied by the widget
+    order: tuple[int, ...]  # A tuple of ints defining the painting order
+    clip: Region  # A region to clip the widget by (if a Widget is within a container)
+    virtual_size: Size  # The virtual size  (scrollable region) of a widget if it is a container
+    container_size: Size  # The container size (area no occupied by scrollbars)
+
+
+RenderRegionMap: TypeAlias = "dict[Widget, RenderRegion]"
+
+
+@rich.repr.auto
+class LayoutUpdate:
+    """A renderable containing the result of a render for a given region."""
+
+    def __init__(self, lines: Lines, region: Region) -> None:
+        self.lines = lines
+        self.region = region
+
+    def __rich_console__(
+        self, console: Console, options: ConsoleOptions
+    ) -> RenderResult:
+        yield Control.home()
+        x = self.region.x
+        new_line = Segment.line()
+        move_to = Control.move_to
+        for last, (y, line) in loop_last(enumerate(self.lines, self.region.y)):
+            yield move_to(x, y)
+            yield from line
+            if not last:
+                yield new_line
+
+    def __rich_repr__(self) -> rich.repr.Result:
+        x, y, width, height = self.region
+        yield "x", x
+        yield "y", y
+        yield "width", width
+        yield "height", height
+
+
+@rich.repr.auto(angular=True)
+class Compositor:
+    """Responsible for storing information regarding the relative positions of Widgets and rendering them."""
+
+    def __init__(self) -> None:
+        # A mapping of Widget on to its "render location" (absolute position / depth)
+        self.map: RenderRegionMap = {}
+
+        # All widgets considered in the arrangement
+        # Not this may be a supperset of self.map.keys() as some widgets may be invisible for various reasons
+        self.widgets: set[Widget] = set()
+
+        # The top level widget
+        self.root: Widget | None = None
+
+        # Dimensions of the arrangement
+        self.size = Size(0, 0)
+
+        # A mapping of Widget on to region, and clip region
+        # The clip region can be considered the window through which a widget is viewed
+        self.regions: dict[Widget, tuple[Region, Region]] = {}
```

Comment:
Initial thought on reading this was: if this is a map of Widget -> Region, and self.map is also a mapping of Widget -> Region, do we need both?

Reply:
You may be right. That information is derived from map anyway. Will see if I can simplify that in a future update.